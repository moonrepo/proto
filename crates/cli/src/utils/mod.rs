pub mod install_graph;
pub mod tool_record;
