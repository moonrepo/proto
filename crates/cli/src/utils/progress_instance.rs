use starbase_console::ConsoleError;
use starbase_console::ui::ProgressReporter;
use std::ops::Deref;
use std::sync::Arc;
use tokio::task::JoinHandle;

pub struct ProgressInstance {
    pub handle: JoinHandle<Result<(), ConsoleError>>,
    pub reporter: Arc<ProgressReporter>,
}

impl ProgressInstance {
    pub async fn stop(self) -> Result<(), ConsoleError> {
        self.reporter.exit();

        if let Ok(result) = self.handle.await {
            result?;
        }

        Ok(())
    }
}

impl Deref for ProgressInstance {
    type Target = ProgressReporter;

    fn deref(&self) -> &Self::Target {
        &self.reporter
    }
}
