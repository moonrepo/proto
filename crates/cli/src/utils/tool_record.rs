use core::ops::{Deref, DerefMut};
use proto_core::{ProtoConfig, ProtoToolConfig, Tool, UnresolvedVersionSpec, VersionSpec};
use std::collections::BTreeMap;

pub struct ToolRecord {
    pub tool: Tool,
    pub config: ProtoToolConfig,
    pub installed_versions: Vec<VersionSpec>,
    pub local_aliases: BTreeMap<String, UnresolvedVersionSpec>,
    pub remote_aliases: BTreeMap<String, UnresolvedVersionSpec>,
    pub remote_versions: Vec<VersionSpec>,
}

impl ToolRecord {
    pub fn new(tool: Tool) -> Self {
        let mut versions = tool
            .inventory
            .manifest
            .installed_versions
            .iter()
            .cloned()
            .collect::<Vec<_>>();
        versions.sort();

        Self {
            tool,
            config: ProtoToolConfig::default(),
            local_aliases: BTreeMap::default(),
            remote_aliases: BTreeMap::default(),
            installed_versions: versions,
            remote_versions: vec![],
        }
    }

    pub fn inherit_from_local(&mut self, config: &ProtoConfig) {
        if let Some(tool_config) = config.tools.get(&self.id).map(|c| c.to_owned()) {
            self.local_aliases.extend(tool_config.aliases.clone());
            self.config = tool_config;
        }
    }

    pub async fn inherit_from_remote(&mut self) -> miette::Result<()> {
        let version_resolver = self
            .tool
            .load_version_resolver(&UnresolvedVersionSpec::default())
            .await?;

        self.remote_aliases.extend(version_resolver.aliases);
        self.remote_versions.extend(version_resolver.versions);
        self.remote_versions.sort();

        Ok(())
    }
}

impl Deref for ToolRecord {
    type Target = Tool;

    fn deref(&self) -> &Self::Target {
        &self.tool
    }
}

impl DerefMut for ToolRecord {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.tool
    }
}
