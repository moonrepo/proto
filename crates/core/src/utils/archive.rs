use crate::helpers::extract_filename_from_url;
use proto_pdk_api::ArchiveSource;
use starbase_archive::{ArchiveError, Archiver};
use starbase_utils::fs::FsError;
use starbase_utils::net::NetError;
use starbase_utils::{fs, net};
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug, miette::Diagnostic)]
pub enum ProtoArchiveError {
    #[diagnostic(transparent)]
    #[error(transparent)]
    Archive(#[from] Box<ArchiveError>),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Fs(#[from] Box<FsError>),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Net(#[from] Box<NetError>),
}

impl From<ArchiveError> for ProtoArchiveError {
    fn from(e: ArchiveError) -> ProtoArchiveError {
        ProtoArchiveError::Archive(Box::new(e))
    }
}

impl From<FsError> for ProtoArchiveError {
    fn from(e: FsError) -> ProtoArchiveError {
        ProtoArchiveError::Fs(Box::new(e))
    }
}

impl From<NetError> for ProtoArchiveError {
    fn from(e: NetError) -> ProtoArchiveError {
        ProtoArchiveError::Net(Box::new(e))
    }
}

pub fn should_unpack(src: &ArchiveSource, target_dir: &Path) -> Result<bool, ProtoArchiveError> {
    let url_file = target_dir.join(".archive-url");
    let mut unpack = true;

    // If the URLs have changed at some point, we need to remove
    // the current files, and download new ones
    if url_file.exists() {
        let previous_url = fs::read_file(&url_file)?;

        if src.url.trim() == previous_url.trim() {
            unpack = false;
        } else {
            fs::remove_dir_all(target_dir)?;
        }
    }

    fs::create_dir_all(target_dir)?;

    Ok(unpack)
}

pub async fn download(
    src: &ArchiveSource,
    temp_dir: &Path,
    client: &reqwest::Client,
) -> Result<PathBuf, ProtoArchiveError> {
    let filename = extract_filename_from_url(&src.url);
    let archive_file = temp_dir.join(&filename);

    net::download_from_url_with_client(&src.url, &archive_file, client).await?;

    Ok(archive_file)
}

pub fn unpack(
    src: &ArchiveSource,
    target_dir: &Path,
    archive_file: &Path,
) -> Result<(String, PathBuf), ProtoArchiveError> {
    let result = unpack_raw(target_dir, archive_file, src.prefix.as_deref());

    fs::write_file(target_dir.join(".archive-url"), &src.url)?;

    result
}

pub fn unpack_raw(
    target_dir: &Path,
    archive_file: &Path,
    prefix: Option<&str>,
) -> Result<(String, PathBuf), ProtoArchiveError> {
    let mut archiver = Archiver::new(target_dir, archive_file);

    if let Some(prefix) = prefix {
        archiver.set_prefix(prefix);
    }

    Ok(archiver.unpack_from_ext()?)
}

pub async fn download_and_unpack(
    src: &ArchiveSource,
    target_dir: &Path,
    temp_dir: &Path,
    client: &reqwest::Client,
) -> Result<(), ProtoArchiveError> {
    if should_unpack(src, target_dir)? {
        let archive_file = download(src, temp_dir, client).await?;

        unpack(src, target_dir, &archive_file)?;
    }

    Ok(())
}
