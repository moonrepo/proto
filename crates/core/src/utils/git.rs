use super::process::{exec_command_piped, handle_exec};
use proto_pdk_api::GitSource;
use starbase_utils::fs;
use std::path::Path;
use tokio::process::Command;

pub fn new_clone(git: &GitSource, cwd: &Path) -> Command {
    let mut cmd = Command::new("git");
    cmd.args(if git.submodules {
        vec!["clone", "--recurse-submodules"]
    } else {
        vec!["clone"]
    })
    .args(["--depth", "1"])
    .arg(&git.url)
    .arg(".")
    .current_dir(cwd);
    cmd
}

pub fn new_checkout(reference: &str, cwd: &Path) -> Command {
    let mut cmd = Command::new("git");
    cmd.arg("checkout").arg(reference).current_dir(cwd);
    cmd
}

pub fn new_pull(cwd: &Path) -> Command {
    let mut cmd = Command::new("git");
    cmd.args(["pull", "--ff", "--prune"]).current_dir(cwd);
    cmd
}

pub async fn clone_or_pull_repo(src: &GitSource, target_dir: &Path) -> miette::Result<()> {
    fs::create_dir_all(target_dir)?;

    if target_dir.join(".git").exists() {
        handle_exec(exec_command_piped(&mut new_pull(target_dir)).await?)?;
    } else {
        handle_exec(exec_command_piped(&mut new_clone(src, target_dir)).await?)?;

        if let Some(reference) = &src.reference {
            handle_exec(exec_command_piped(&mut new_checkout(reference, target_dir)).await?)?;
        }
    }

    Ok(())
}
