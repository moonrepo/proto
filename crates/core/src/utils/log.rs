use starbase_utils::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

#[derive(Clone, Default)]
pub struct LogWriter {
    buffer: Arc<Mutex<Vec<String>>>,
}

impl LogWriter {
    fn add_title_row(&self, depth: u8, title: impl AsRef<str>) {
        let mut buffer = self.buffer.lock().unwrap();
        buffer.push(format!("{} {}", "#".repeat(depth as usize), title.as_ref()));
        buffer.push("".into());
    }

    pub fn add_title(&self, title: impl AsRef<str>) {
        self.add_title_row(1, title);
    }

    pub fn add_header(&self, title: impl AsRef<str>) {
        self.add_title_row(2, title);
    }

    pub fn add_section(&self, title: impl AsRef<str>) {
        self.add_title_row(3, title);
    }

    pub fn add_subsection(&self, title: impl AsRef<str>) {
        self.add_title_row(4, title);
    }

    pub fn add_code(&self, label: impl AsRef<str>, value: impl AsRef<str>) {
        let mut buffer = self.buffer.lock().unwrap();
        let value = value.as_ref().trim();

        buffer.push(format!("**{}**:", label.as_ref().to_uppercase()));

        if !value.is_empty() {
            buffer.push("```".into());
            buffer.push(value.into());
            buffer.push("```".into());
        }

        buffer.push("".into());
    }

    pub fn add_error(&self, error: impl AsRef<dyn std::error::Error>) {
        let error = error.as_ref();
        let ansi = regex::Regex::new(r"\x1b\[([\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])").unwrap();

        self.add_code("ERROR", ansi.replace_all(&error.to_string(), ""));

        let mut source = error.source();

        while let Some(cause) = source {
            self.add_code("CAUSED BY", ansi.replace_all(&cause.to_string(), ""));
            source = cause.source();
        }
    }

    // pub fn add_line(&self) {
    //     let mut buffer = self.buffer.lock().unwrap();
    //     buffer.push("".into());
    // }

    pub fn add_value(&self, label: impl AsRef<str>, value: impl AsRef<str>) {
        let mut buffer = self.buffer.lock().unwrap();
        buffer.push(format!(
            "**{}**: {}",
            label.as_ref().to_uppercase(),
            value.as_ref()
        ));
        buffer.push("".into());
    }

    pub fn add_value_opt<T: AsRef<str>>(&self, label: impl AsRef<str>, value: Option<T>) {
        if let Some(value) = value {
            self.add_value(label, value);
        }
    }

    pub fn write_to(&self, path: PathBuf) -> miette::Result<()> {
        let mut buffer = self.buffer.lock().unwrap();

        if !buffer.is_empty() {
            buffer.push("".into());

            fs::write_file(path, buffer.join("\n"))?;
        }

        Ok(())
    }
}
