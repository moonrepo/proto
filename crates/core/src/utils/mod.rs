pub mod archive;
pub mod git;
pub mod process;
