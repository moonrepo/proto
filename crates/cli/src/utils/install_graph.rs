use super::tool_record::ToolRecord;
use proto_core::Id;
use rustc_hash::{FxHashMap, FxHashSet};
use std::sync::Arc;
use tokio::sync::RwLock;

pub enum InstallStatus {
    ReqFailed(Id),
    WaitingOnReqs(Vec<Id>),
}

#[derive(Clone)]
pub struct InstallGraph {
    /// Tools that must be installed
    ids: Arc<FxHashSet<Id>>,

    /// Tools that have been installed
    installed: Arc<RwLock<FxHashSet<Id>>>,

    /// Tools that have been not installed (some error occurred)
    not_installed: Arc<RwLock<FxHashSet<Id>>>,

    /// Tools that require other tools to be installed
    /// before they can be installed
    requires: Arc<FxHashMap<Id, Vec<Id>>>,
}

impl InstallGraph {
    pub fn new(tools: &[ToolRecord]) -> Self {
        let mut ids = FxHashSet::default();
        let mut requires = FxHashMap::default();

        for tool in tools {
            ids.insert(tool.id.clone());

            if !tool.metadata.requires.is_empty() {
                requires.insert(
                    tool.id.clone(),
                    tool.metadata.requires.iter().map(Id::raw).collect(),
                );
            }
        }

        Self {
            ids: Arc::new(ids),
            installed: Arc::new(RwLock::new(FxHashSet::default())),
            not_installed: Arc::new(RwLock::new(FxHashSet::default())),
            requires: Arc::new(requires),
        }
    }

    pub async fn check_install_status(&self, id: &Id) -> Option<InstallStatus> {
        if !self.ids.contains(id) {
            return None;
        }

        if let Some(require_ids) = self.requires.get(id) {
            let installed = self.installed.read().await;
            let not_installed = self.not_installed.read().await;
            let mut waiting_on = vec![];

            for require_id in require_ids {
                if !self.ids.contains(require_id) {
                    continue;
                }

                if !installed.contains(require_id) {
                    waiting_on.push(require_id.clone());
                } else if not_installed.contains(require_id) {
                    return Some(InstallStatus::ReqFailed(require_id.clone()));
                }
            }

            if !waiting_on.is_empty() {
                return Some(InstallStatus::WaitingOnReqs(waiting_on));
            }
        }

        None
    }

    pub async fn mark_installed(&self, id: &Id) {
        self.installed.write().await.insert(id.to_owned());
    }

    pub async fn mark_not_installed(&self, id: &Id) {
        self.not_installed.write().await.insert(id.to_owned());
    }
}
