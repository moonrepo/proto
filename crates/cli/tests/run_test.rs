mod utils;

use starbase_sandbox::{assert_snapshot, predicates::prelude::*, Sandbox};
use std::{env, fs};
use utils::*;

fn install_node(sandbox: &Sandbox) {
    sandbox
        .run_bin(|cmd| {
            cmd.arg("install").arg("node").arg("19.0.0");
        })
        .success();
}

mod run {
    use super::*;

    #[test]
    fn errors_if_not_installed() {
        let sandbox = create_empty_proto_sandbox();

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run").arg("node").arg("19.0.0");
            })
            .failure();

        assert.stderr(predicate::str::contains(
            "This project requires Node.js 19.0.0",
        ));
    }

    #[test]
    fn errors_if_no_version_detected() {
        let sandbox = create_empty_proto_sandbox();

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run").arg("node");
            })
            .failure();

        assert.stderr(predicate::str::contains(
            "Failed to detect an applicable version",
        ));
    }

    #[test]
    fn runs_a_tool() {
        let sandbox = create_empty_proto_sandbox();

        sandbox
            .run_bin(|cmd| {
                cmd.arg("install").arg("node").arg("19.0.0");
            })
            .success();

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("19.0.0"));
    }

    #[test]
    fn runs_a_tool_using_version_detection() {
        let sandbox = create_empty_proto_sandbox();

        sandbox
            .run_bin(|cmd| {
                cmd.arg("install").arg("node").arg("19.0.0");
            })
            .success();

        // Arg
        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("19.0.0"));

        // Env var
        let assert = sandbox
            .run_bin(|cmd| {
                cmd.env("PROTO_NODE_VERSION", "19.0.0")
                    .arg("run")
                    .arg("node")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("19.0.0"));

        // Local version
        sandbox.create_file(".prototools", "node = \"19.0.0\"");

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run").arg("node").arg("--").arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("19.0.0"));

        fs::remove_file(sandbox.path().join(".prototools")).unwrap();

        // Global version
        sandbox.create_file(".proto/.prototools", "node = \"19.0.0\"");

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run").arg("node").arg("--").arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("19.0.0"));
    }

    #[test]
    fn updates_last_used_at() {
        let sandbox = create_empty_proto_sandbox();
        let last_used_file = sandbox.path().join(".proto/tools/node/19.0.0/.last-used");

        sandbox
            .run_bin(|cmd| {
                cmd.arg("install").arg("node").arg("19.0.0");
            })
            .success();

        assert!(!last_used_file.exists());

        sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        let value = fs::read_to_string(&last_used_file).unwrap();

        assert!(last_used_file.exists());
        assert_ne!(value, "");

        // Run again and make sure timestamps update
        sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        let new_value = fs::read_to_string(&last_used_file).unwrap();

        assert!(last_used_file.exists());
        assert_ne!(value, new_value);
    }

    #[test]
    fn auto_installs_if_missing() {
        let sandbox = create_empty_proto_sandbox();

        sandbox.create_file(".prototools", "[settings]\nauto-install = true");

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("Node.js 19.0.0 installed"));
    }

    #[test]
    fn auto_installs_if_missing_with_env_var() {
        let sandbox = create_empty_proto_sandbox();

        env::set_var("PROTO_AUTO_INSTALL", "true");

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("Node.js 19.0.0 installed"));

        env::remove_var("PROTO_AUTO_INSTALL");
    }

    #[test]
    fn doesnt_auto_install_if_false() {
        let sandbox = create_empty_proto_sandbox();

        sandbox.create_file(".prototools", "[settings]\nauto-install = false");

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run").arg("node").arg("19.0.0");
            })
            .failure();

        assert.stderr(predicate::str::contains(
            "This project requires Node.js 19.0.0",
        ));
    }

    #[test]
    fn doesnt_auto_install_subsequently() {
        let sandbox = create_empty_proto_sandbox();

        sandbox.create_file(".prototools", "[settings]\nauto-install = true");

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("Node.js 19.0.0 installed"));

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("--version");
            })
            .success();

        assert.stdout(predicate::str::contains("Node.js 19.0.0 installed").not());
    }

    #[test]
    fn errors_if_plugin_not_configured() {
        let sandbox = create_empty_proto_sandbox();

        let assert = sandbox
            .run_bin(|cmd| {
                cmd.arg("run").arg("plugin-name").arg("1.0.0");
            })
            .failure();

        assert.stderr(predicate::str::contains(
            "plugin-name is not a built-in tool",
        ));
    }

    mod env_vars {
        use super::*;

        #[test]
        fn inherits_from_config() {
            let sandbox = create_proto_sandbox("env-vars");

            sandbox.create_file(
                ".prototools",
                r#"
[tools.node.env]
FROM_CONFIG = "abc123"
FROM_CONFIG_BOOL = true
"#,
            );

            install_node(&sandbox);

            let assert = sandbox.run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("test.js");
            });

            assert_snapshot!(assert.output_standardized());
        }

        #[test]
        fn inherits_from_parent() {
            let sandbox = create_proto_sandbox("env-vars");

            install_node(&sandbox);

            let assert = sandbox.run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("test.js")
                    .env("FROM_PARENT", "abc123");
            });

            assert_snapshot!(assert.output_standardized());
        }

        #[test]
        fn can_disable_inherits_from_parent_with_config() {
            let sandbox = create_proto_sandbox("env-vars");

            sandbox.create_file(
                ".prototools",
                r#"
[tools.node.env]
FROM_PARENT_REMOVED = false
"#,
            );

            install_node(&sandbox);

            let assert = sandbox.run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("test.js")
                    .env("FROM_PARENT", "abc123")
                    .env("FROM_PARENT_REMOVED", "abc123");
            });

            assert_snapshot!(assert.output_standardized());
        }

        #[test]
        fn parent_overrides_config() {
            let sandbox = create_proto_sandbox("env-vars");

            sandbox.create_file(
                ".prototools",
                r#"
[tools.node.env]
FROM_CONFIG = "abc123"
"#,
            );

            install_node(&sandbox);

            let assert = sandbox.run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("test.js")
                    .env("FROM_CONFIG", "xyz789")
                    .env("FROM_PARENT", "xyz789");
            });

            assert_snapshot!(assert.output_standardized());
        }

        #[test]
        fn supports_interpolation() {
            let sandbox = create_proto_sandbox("env-vars");

            sandbox.create_file(
                ".prototools",
                r#"
[tools.node.env]
FIRST = "abc"
SECOND = "123"
THIRD = "value-${FIRST}-${SECOND}-${PARENT}"
FOURTH = "ignores-$FIRST-$PARENT"
"#,
            );

            install_node(&sandbox);

            let assert = sandbox.run_bin(|cmd| {
                cmd.arg("run")
                    .arg("node")
                    .arg("19.0.0")
                    .arg("--")
                    .arg("interpolation.js")
                    .env("SECOND", "789")
                    .env("PARENT", "xyz");
            });

            assert_snapshot!(assert.output_standardized());
        }
    }
}
