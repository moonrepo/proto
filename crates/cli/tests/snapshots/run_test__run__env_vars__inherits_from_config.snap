---
source: crates/cli/tests/run_test.rs
expression: get_assert_output(&assert)
---
FROM_CONFIG = abc123
FROM_CONFIG_BOOL = true
FROM_PARENT = undefined
FROM_PARENT_REMOVED = undefined


